Redis performance improvements

brasic opened the pull request:
This contains three optimizations I noticed while reviewing this gem:
 - `evalsha` should be preferred to `eval`, since it reduces script compilation and network overhead (https://github.com/ShippingEasy/gcra-ruby/commit/490b7be28138cbf207828b324646648263e68601)
 - independent adjacent calls to redis should be pipelined to avoid multiple round trips (https://github.com/ShippingEasy/gcra-ruby/commit/035d6ed40a52a38aa74bb18c011ee85012ee4a2b)
 - A combination of `SETNX` and `PEXPIRE` is equivalent to and can be replaced with a single `SET` call with `NX` and `PX` options (https://github.com/ShippingEasy/gcra-ruby/commit/722c516cdc01b067a3ba4083be0e1207fbb0587f)

brasic committed:
Use evalsha rather than eval

Passing the full script source to redis every time is expensive in redis
compute time and network bandwidth.  loading the script once per store
and calling `evalsha` is faster.  The only tricky part is what to do if
the script cache is tampered with; this requires reloading the script
and a single retry in `compare_and_set_with_ttl`.

brasic committed:
Pipeline adjacent redis calls

No reason this set of two calls has to be two network roundtrips

brasic committed:
Make set_if_not_exists_with_ttl atomic and single-operation

The SET command was enhanced in redis 2.6 to support NX and PX options,
which allow us to implement "set if not exists" + "expire in ms"
functionality in a single command by passing the appropriate options.

This gem already requires redis 2.6 so there is no reason not to use
SET in this way rather than SETNX+PEXPIRE

christopherbradbury commented on lib/gcra/redis_store.rb:
```diff
@@ -22,10 +23,12 @@ def initialize(redis, key_prefix)
     # Also returns the time from the Redis server, with microsecond precision.
     def get_with_time(key)
-      time_response = @redis.time # returns tuple (seconds since epoch, microseconds)
+      time_response, value = @redis.pipelined do
```
curious, what is the benefit of setting value with `@redis.pipelined` like this? less calls to key store?

brasic replied:
See https://redis.io/topics/pipelining.  This sends both commands in a single network payload and then decodes both responses.  It reduces the number of network roundtrips, although from the redis server's point of view it is still two commands.  But redis is so fast that the network round trip time always vastly dominates the time taken to execute commands.

joshuaflanagan commented on lib/gcra/redis_store.rb:
```diff
@@ -67,6 +70,10 @@ def compare_and_set_with_ttl(key, old_value, new_value, ttl_nano)
     private
 
+    def cas_sha
+      @cas_sha ||= @redis.script('load', CAS_SCRIPT)
```
I _think_ we could go even further and pre-compute the hash from the script constant and send that by default. On error, call SCRIPT LOAD or EVAL. Currently, this will pass the entire script for each instance of this class. But scripts stay loaded forever - they are not scoped to a single connection.

Not a huge deal, but we are optimizing...

brasic replied:
Fair point!

brasic replied:
addressed in 6474c7aace4a411f7e5159a1a3e490fbb4b1feb9

brasic committed:
Compute SHA1 client-side, only load to redis on NOSCRIPT

This avoids a bunch of clients all calling SCRIPT LOAD with the same
script when booting up.
